### PR TITLE
some new code

### PR DESCRIPTION
**Pull request by 1Evolve**
## Description

Rewrite the camera code
With the addition of new codes, and adjusting the code in a better way It is now easier to add new cameras to the code
Add a list of all cameras with numbers, names and addresses Add a tablet and animation when opening the camera andyy code

## Checklist

- [x] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [x] My pull request fits the contribution guidelines & code conventions.

### CODE REVIEW

**Comment by 1Evolve** [comment]
https://github.com/Qbox-project/qbx_radialmenu/pull/44

**Comment by 1Evolve** [comment]
for add new camera easy using this code 

```
RegisterCommand('draw', function()
    if isInitialCast then return end
    isInitialCast = true
    lib.hideTextUI()
    lib.showTextUI('[E] - Copy Coords', { position = "left-center", })
    CreateThread(function() 
        while isInitialCast do
            local hit, _, hitCoords, _, _ = lib.raycast.cam(1, 4, 10)
            
            if hit then
                local playerHeadCoords = GetPedBoneCoords(cache.ped, 31086, 0, 0, 0)
                DrawLine(playerHeadCoords.x, playerHeadCoords.y, playerHeadCoords.z, hitCoords.x, hitCoords.y, hitCoords.z, 29, 21, 237, 255)
                DrawSphere(hitCoords.x, hitCoords.y, hitCoords.z, 0.05, 29, 21, 237, 0.85)
                if IsControlJustPressed(0, 38) then
                    initialCoords = hitCoords
                    lib.hideTextUI()
                    isInitialCast = false
                    print(initialCoords.x, initialCoords.y, initialCoords.z, initialCoords.w)
                end
            end
        end
    end)
end, false)
```

File: config/client.lua [hunk 1]
```diff
@@ -6,57 +6,56 @@ return {
...
+        [40] = {coords = vec4(265.72, 213.44, 111.15, 97.76), label = 'Pacific Bank - CAM#1', isOnline = true},
+        [41] = {coords = vec4(233.0, 221.46, 106.29, 193.4), label = 'Pacific Bank - CAM#2', isOnline = true},
+        [42] = {coords = vec4(240.29, 230.29, 106.28, 120.92), label = 'Pacific Bank - CAM#3', isOnline = true},
+        [43] = {coords = vec4(-103.92, 6467.74, 31.63, 111.51), label = 'Paleto Bank - CAM#1', isOnline = true},
+        [44] = {coords = vec4(-627.89, -239.76, 38.06, 354.4), label = 'Vangelico Jewelery - CAM#1', isOnline = true},
+        [45] = {coords = vec4(-627.09, -230.15, 38.06, 187.17), label = 'Vangelico Jewelery - CAM#2', isOnline = true},
+        [46] = {coords = vec4(-620.23, -224.39, 38.06, 138.98), label = 'Vangelico Jewelery - CAM#3', isOnline = true},
+        [47] = {coords = vec4(-622.57, -235.57, 38.06, 53.63), label = 'Vangelico Jewelery - CAM#4', isOnline = true},
+        [48] = {coords = vec4(434.190, -978.340, 33.165, 53.63), label = 'Los Santos Police Department - CAM#1', isOnline = true},
+        [49] = {coords = vec4(458.513, -1017.173, 31.428, 53.63), label = 'Los Santos Police Department - CAM#2', isOnline = true},
```
Comment (BerkieBb) [inline config/client.lua:58]: These numbers in front are not necessary, lua adds these automatically and in vscode you can enable in the lua extension to show these as hints instead.

File: client/camera.lua [hunk 2]
```diff
@@ -1,209 +1,171 @@
...
-        if createdCam == 0 or currentScaleform == -1 then
-            Wait(2000)
-        else
-            listenForCamControls()
-            Wait(0)
+        if IsControlJustPressed(1, 177) then -- ESC - Backspace
+            InsideCam = false
         end
+
+        Wait(1)
```
Comment (BerkieBb) [inline client/camera.lua:160]: Never wait 1, remember? Wait(0) for every frame stuff

File: client/camera.lua [hunk 3]
```diff
@@ -1,209 +1,171 @@
...
     end
+
+    lib.hideTextUI()
+    removeTablet(cache.ped)
+    ClearPedTasks(PlayerPedId())
```
Comment (BerkieBb) [inline client/camera.lua:165]: cache.ped can be used here

File: client/camera.lua [hunk 4]
```diff
@@ -1,209 +1,171 @@
...
+RegisterNetEvent('police:client:showcamera', function()
+    local menu = {
+        id = 'camera_menu',
+        title = 'Camera List',
+        options = {}
+    }
+
+    for camId, cameraData in pairs(config.securityCameras) do
+        if cameraData.isOnline then
+            table.insert(menu.options, {
```
Comment (BerkieBb) [inline client/camera.lua:78]: Use `menu.options[#menu.options + 1] =` instead of table.insert

File: client/camera.lua [hunk 5]
```diff
@@ -1,209 +1,171 @@
...
+        end
+    end
+
+    lib.registerContext(menu)
+    lib.showContext('camera_menu')
+end)
+
+RegisterNetEvent('police:client:opencamera', function(cameraId)
+    local coords = config.securityCameras[tonumber(cameraId)].coords
+    cameraId = tonumber(cameraId)
```
Comment (BerkieBb) [inline client/camera.lua:93]: Put this above the local coords so you don't tonumber twice

File: client/camera.lua [hunk 6]
```diff
@@ -1,209 +1,171 @@
...
+    useTablet(cache.ped)
+
+    while InsideCam do
+        SetCamCoord(securityCam, coords.x, coords.y, (coords.z + 1.2))
+        PushScaleformMovieFunction(scaleform, 'SET_ALT_FOV_HEADING')
+        PushScaleformMovieFunctionParameterFloat(1.0)
+        PushScaleformMovieFunctionParameterFloat(GetCamRot(securityCam, 2).z)
+        PopScaleformMovieFunctionVoid()
+        DrawScaleformMovieFullscreen(scaleform, 255, 255, 255, 255)
+        local CamRot = GetCamRot(securityCam, 2)
```
Comment (BerkieBb) [inline client/camera.lua:123]: camelCase

File: client/camera.lua [hunk 7]
```diff
@@ -1,209 +1,171 @@
...
-    PushScaleformMovieFunction(scaleform, 'SET_BACKGROUND_COLOUR')
-    PushScaleformMovieFunctionParameterInt(0)
-    PushScaleformMovieFunctionParameterInt(0)
-    PushScaleformMovieFunctionParameterInt(0)
-    PushScaleformMovieFunctionParameterInt(80)
-    PopScaleformMovieFunctionVoid()
-
-    return scaleform
+local config = require 'config.client'
+local InsideCam = false
```
Comment (BerkieBb) [inline client/camera.lua:2]: camelCase

File: server/commands.lua [hunk 8]
```diff
@@ -184,18 +184,9 @@ lib.addCommand('sc', {help = Lang:t('commands.softcuff')}, function(source)
...
-            name = 'camid',
-            type = 'number',
-            help = Lang:t('info.camera_id_help')
-        }
-    },
- }, function(source, args)
+lib.addCommand({'cam', 'camera'}, {help = Lang:t('commands.camera')}, function(source)
     if not checkLeoAndOnDuty(source) then return end
-    TriggerClientEvent('police:client:ActiveCamera', source, args.camid)
+    TriggerClientEvent('police:client:showcamera', source)
```
Comment (BerkieBb) [inline server/commands.lua:189]: camelCase

**Review by BerkieBb — CHANGES_REQUESTED** [review]

File: server/commands.lua (hunk 8 above)
Comment (1Evolve) [reply server/commands.lua]: done :)

File: client/camera.lua (hunk 7 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

File: client/camera.lua (hunk 6 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

File: client/camera.lua (hunk 5 above)
Comment (1Evolve) [reply client/camera.lua:44]: done :)

File: client/camera.lua (hunk 4 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

File: client/camera.lua (hunk 3 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

**Comment by BerkieBb** [comment]
Check remaining comments

File: client/camera.lua (hunk 2 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

File: config/client.lua (hunk 1 above)
Comment (1Evolve) [reply config/client.lua]: yes remove number now

**Comment by 1Evolve** [comment]
Is there any response to the request?

**Comment by 1Evolve** [comment]
If it does not fit the idea of the project, please close the request

**Comment by 1Evolve** [comment]
three months

File: server/commands.lua [hunk 9]
```diff
@@ -184,18 +184,9 @@ lib.addCommand('sc', {help = Lang:t('commands.softcuff')}, function(source)
...
-            name = 'camid',
-            type = 'number',
-            help = Lang:t('info.camera_id_help')
-        }
-    },
- }, function(source, args)
+lib.addCommand({'cam', 'camera'}, {help = Lang:t('commands.camera')}, function(source)
     if not checkLeoAndOnDuty(source) then return end
-    TriggerClientEvent('police:client:ActiveCamera', source, args.camid)
+    TriggerClientEvent('police:client:showCamera', source)
```
Comment (Manason) [inline server/commands.lua:189]: Would this be better as a callback? I like to use events only when we expect/want multiple event handlers, but this seems to be more like a network function call use of events.

File: client/camera.lua [hunk 10]
```diff
@@ -1,209 +1,171 @@
...
+local function useTablet(ped)
+    lib.requestModel(`prop_cs_tablet`)
+    lib.requestAnimDict('amb@code_human_in_bus_passenger_idles@female@tablet@base')
+
+    if DoesEntityExist(tabletProp) then
+        DeleteEntity(tabletProp)
     end
-    return tostring(hours .. ':' .. minutes)
+    
+    tabletProp = CreateObject(`prop_cs_tablet`, 0.0, 0.0, 0.0, true, true, false)
```
Comment (Manason) [inline client/camera.lua:27]: We shouldn't be creating networked props on the client as it is not compliant with entity lockdown mode. Instead, create it on the server.

File: client/camera.lua [hunk 11]
```diff
@@ -1,209 +1,171 @@
...
-    PopScaleformMovieFunctionVoid()
+    SetTimecycleModifier('heliGunCam')
+    SetTimecycleModifierStrength(1.0)
 
-    PushScaleformMovieFunction(scaleform, 'DRAW_INSTRUCTIONAL_BUTTONS')
-    PopScaleformMovieFunctionVoid()
+    local scaleform = RequestScaleformMovie('TRAFFIC_CAM')
+    while not HasScaleformMovieLoaded(scaleform) do
+        Wait(0)
+    end
```
Comment (Manason) [inline client/camera.lua:53]: Use https://overextended.dev/ox_lib/Modules/Streaming/Client#librequestscaleformmovie

File: client/job.lua [hunk 12]
```diff
@@ -10,6 +10,11 @@ local inImpound = false
 local inGarage = false
 local inPrompt = false
 
+local function isLeo(level, onduty)
+    local playerJob = QBX.PlayerData.job
+    return playerJob.type ~= 'leo' or playerJob.grade.level >= (level or 0) or playerJob.onduty == (onduty == nil and true or onduty)
```
Comment (Manason) [inline client/job.lua:15]:
```suggestion
    return playerJob.type == 'leo' and playerJob.grade.level >= (level or 0) and playerJob.onduty == (onduty == nil and true or onduty)
```

**Review by Manason — CHANGES_REQUESTED** [review]
Please fix the lint errors.

It looks like there are several new features and refactors in this PR. If you want faster reviews, consider breaking up your PRs into smaller bite sized chunks of one new feature, or one refactor. These large PRs can be difficult and time consuming to review with lots of back and forth comments.

File: client/camera.lua (hunk 11 above)
Comment (1Evolve) [reply client/camera.lua]: done :)

File: client/job.lua (hunk 12 above)
Comment (1Evolve) [reply client/job.lua]: thank you for edit :)

File: client/main.lua [hunk 13]
```diff
@@ -7,18 +8,42 @@ local dutyBlips = {}
...
+        elseif classVeh == 18 then -- walk
+            SetBlipSprite(blip, 56)
+            ShowHeadingIndicatorOnBlip(blip, true)
+        elseif classVeh == 0 then
+            SetBlipSprite(blip, 126)
+            ShowHeadingIndicatorOnBlip(blip, true)
+        else -- all vehicles
+            SetBlipSprite(blip, 225)
+            ShowHeadingIndicatorOnBlip(blip, true)
+        end
```
[the remark this answers is not in the document]
Comment (1Evolve) [reply client/main.lua]: I think it works better now thank u for sugg

File: server/commands.lua (hunk 9 above)
Comment (BerkieBb) [reply server/commands.lua:189]: A callback is only used when you expect to receive something back, otherwise you use events. So this is correct.

Comment (BerkieBb) [reply server/commands.lua:189]: Callback inherently are also just events, so it won't make a difference if you do use it aside from you using more resources to do the same thing

File: client/camera.lua (hunk 10 above)
Comment (BerkieBb) [reply client/camera.lua:27]: @1Evolve just this one and you should be good

**Comment by 1Evolve** [comment]
Completed 4 months ago
You can delete the withdrawal request
I'm no longer as enthusiastic as I started out with

**Comment by mafewtm** [comment]
Closed as requested.

**Pull request closed without merging**